albatross/core/model.h

```diff
@@ -13,266 +13,110 @@
-    std::ostringstream ss;
-    ss << get_name() << std::endl;
-    ss << ParameterHandlingMixin::pretty_string();
-    return ss.str();
-  }
+  template <typename X, typename Y, typename Z>
+  friend class Prediction;
 
-  virtual bool has_been_fit() const { return has_been_fit_; }
+  template <typename T, typename FeatureType>
+  friend class fit_model_type;
 
-  virtual std::string get_name() const = 0;
+  template <typename T, typename FitModelType>
+  friend struct fit_type_from_fit_model_type;
 
-  virtual Insights get_insights() const { return insights_; }
+  template <typename T, typename FeatureType>
+  friend struct fit_type;
 
-  virtual void add_insights(const Insights &insights) {
-    for (const auto &insight : insights) {
-      insights_[insight.first] = insight.second;
-    }
-  };
-
-  virtual std::unique_ptr<RegressionModel<FeatureType>>
-  ransac_model(double inlier_threshold, std::size_t min_inliers,
-               std::size_t random_sample_size, std::size_t max_iterations) {
-    static_assert(
-        is_complete<GenericRansac<void, FeatureType>>::value,
-        "ransac methods aren't complete yet, be sure you've included ransac.h");
-    return make_generic_ransac_model<FeatureType>(
-        this, inlier_threshold, min_inliers, random_sample_size, max_iterations,
-        leave_one_out_indexer<FeatureType>);
-  }
+ private:
+  // Declaring these private makes it impossible to accidentally do things like:
+  //     class A : public ModelBase<B> {}
+  // or
+  //     using A = ModelBase<B>;
+  //
+  // which if unchecked can lead to some very strange behavior.
+  ModelBase() : insights_(){};
+  friend ModelType;
+  Insights insights_;
 
   /*
-   * Here we define the serialization routines.  Note that while in most
-   * cases we could use the cereal method `serialize`, in this case we don't
-   * know for sure where the parameters are stored.  The
-   * GaussianProcessRegression
-   * model, for example, derives its parameters from its covariance function,
-   * so it's `params_` are actually empty.  As a result we need to use the
-   * save/load cereal variant and deal with parameters through the get/set
-   * interface.
+   * Fit
    */
-  template <class Archive> void save(Archive &archive) const {
-    auto params = get_params();
-    archive(cereal::make_nvp("parameters", params));
-    archive(cereal::make_nvp("has_been_fit", has_been_fit_));
-  }
-
-  template <class Archive> void load(Archive &archive) {
-    auto params = get_params();
-    archive(cereal::make_nvp("parameters", params));
-    archive(cereal::make_nvp("has_been_fit", has_been_fit_));
-    set_params(params);
-  }
+  template <
+      typename FeatureType,
+      typename std::enable_if<has_valid_fit<ModelType, FeatureType>::value,
+                              int>::type = 0>
+  auto
+  fit_(const std::vector<FeatureType> &features,
+       const MarginalDistribution &targets) const {
+    const auto fit = derived().fit(features, targets);
```

Comment:
`fit` can't be `const` if you're going to be moving it into the `FitModel` constructor below.